libplatsupport/src/mach/imx/serial/serial.c

```diff
@@ -102,18 +102,34 @@ int uart_getchar(ps_chardevice_t *d)
     return c;
 }
 
+static int
+internal_uart_tx_busy(imx_uart_regs_t* regs)
+{
+    return !(regs->sr2 & UART_SR2_TXFIFO_EMPTY);
+}
+
+static int
+internal_uart_tx(imx_uart_regs_t* regs, int c)
+{
+    regs->txd = c;
+}
+
 int uart_putchar(ps_chardevice_t* d, int c)
 {
     imx_uart_regs_t* regs = imx_uart_get_priv(d);
-    if (regs->sr2 & UART_SR2_TXFIFO_EMPTY) {
-        if (c == '\n' && (d->flags & SERIAL_AUTO_CR)) {
-            uart_putchar(d, '\r');
-        }
-        regs->txd = c;
-        return c;
-    } else {
+    if (internal_uart_tx_busy(regs)) {
         return -1;
     }
+
+    if (c == '\n' && (d->flags & SERIAL_AUTO_CR)) {
+        internal_uart_tx(regs, '\r');
+        if (internal_uart_tx_busy(regs)) {
```

Comment:
This pattern causes an infinite loop where internal_uart_tx_busy returns true immediately after internal_uart_tx.  This means that when a '\n' is printed it turns into an infinite stream of '\r' chars.  This is observed on the imx31, imx6, tk1, tx1 and tx2 platforms.

Reply:
thanks, will look into this